File: radix_sort.py
"""
File implementing radix sort algorithm using Python
Name: Marcela Perro 
Semester: Spring 2024 
"""

import time
import random
import sys

MAX = 10  # Assuming we are dealing with decimal numbers

# Simulated memory allocation tracking
allocated_memory = 0
"""
    Tracks memory allocation for simulation purposes.
    Adds the specified size to the global allocated_memory counter.
"""
def memory_track(size):
    global allocated_memory
    allocated_memory += size

"""
    Returns the maximum value in the given array.
"""
def getMax(arr):
    return max(arr)

"""
    Performs counting sort on the array 'arr' based on the digit at the exponent 'exp'.
"""
def countingSort(arr, exp):
    n = len(arr)
    output = [0] * n #Output array initialized with zeros 
    count = [0] * MAX #Count array for storng digit frequencies 
    memory_track(sys.getsizeof(output) + sys.getsizeof(count))

    #Store count of occurrences of each digit 
    for i in arr:
        index = i // exp
        count[index % 10] += 1

    #Change count index to contain the actual position of this digit in output 
    for i in range(1, MAX):
        count[i] += count[i - 1]

    #Build the output array 
    i = n - 1
    while i >= 0:
        index = arr[i] // exp
        output[count[index % 10] - 1] = arr[i]
        count[index % 10] -= 1
        i -= 1

    #Copy the sorted elemetns back into the original array 
    for i in range(n):
        arr[i] = output[i]

"""
    Main function to sort an array using the Radix Sort algorithm.
"""
def radixSort(arr):
    m = getMax(arr)
    exp = 1
    while m // exp > 0:
        countingSort(arr, exp)
        exp *= 10

"""
    Utility function to print the elements of the array.
"""
def print_array(arr):
    print(' '.join(str(i) for i in arr))

"""
    Main function that executes the Radix Sort for arrays of different sizes and logs the results.
"""
def main(max_n, output_file='radix_sort_results.csv'):
    with open(output_file, 'w') as file:
        file.write("Array Size,Time Taken (seconds),Memory Used (bytes)\n")

        for n in range(1, max_n + 1):
            global allocated_memory
            allocated_memory = 0

            arr = [random.randint(0, 999) for _ in range(n)]
            memory_track(sys.getsizeof(arr))

            start_time = time.time()
            radixSort(arr)
            elapsed_time = time.time() - start_time

            print(f"Array Size: {n}, Time Taken: {elapsed_time:.6f} seconds, Memory Used: {allocated_memory} bytes")
            file.write(f"{n},{elapsed_time:.6f},{allocated_memory}\n")

if __name__ == "__main__":
    if len(sys.argv) < 2:
        print("Usage: python3 radix_sort.py <max array size> [output file]")
        sys.exit(1)

    max_n = int(sys.argv[1])
    output_file = sys.argv[2] if len(sys.argv) > 2 else "radix_sort_results.csv"
    main(max_n, output_file)

File: radix_sort.c
/*
File implementing radix sort algorithm using C 
Name: Marcela Perro 
Semester: Spring 2024 
*/
#include <stdio.h>
#include <stdlib.h>
#include <time.h>
#include <stdbool.h>
#include <string.h>


#define MAX 10  // Assuming we are dealing with decimal numbers

// Global variable to track memory usage
size_t allocated_memory = 0;

// Custom memory allocation function
void* malloc_track(size_t size) {
    void *ptr = malloc(size);
    if (ptr) {
        allocated_memory += size;
    }
    return ptr;
}

// Custom memory deallocation function
void free_track(void *ptr) {
    if (ptr) {
        //allocated_memory -= size;
        free(ptr);
    }
}

// A utility function to get the maximum value in arr[]
int getMax(int arr[], int n) {
    int mx = arr[0];
    for (int i = 1; i < n; i++)
        if (arr[i] > mx)
            mx = arr[i];
    return mx;
}

// A function to do counting sort of arr[] according to the digit represented by exp.
void countingSort(int arr[], int n, int exp) {
    int output[n];
    int i, count[MAX] = {0};

    // Store count of occurrences in count[]
    for (i = 0; i < n; i++)
        count[(arr[i] / exp) % 10]++;

    // Change count[i] so that count[i] now contains actual position of this digit in output[]
    for (i = 1; i < MAX; i++)
        count[i] += count[i - 1];

    // Build the output array
    for (i = n - 1; i >= 0; i--) {
        output[count[(arr[i] / exp) % 10] - 1] = arr[i];
        count[(arr[i] / exp) % 10]--;
    }

    // Copy the output array to arr[], so that arr[] now contains sorted numbers according to current digit
    for (i = 0; i < n; i++)
        arr[i] = output[i];
}

// The main function to that sorts arr[] of size n using Radix Sort
void radixSort(int arr[], int n) {
    int m = getMax(arr, n);

    for (int exp = 1; m / exp > 0; exp *= 10) {
        countingSort(arr, n, exp);
    }
}

// A utility function to print an array
void print_array(int arr[], int n) {
    for (int i = 0; i < n; i++)
        printf("%d ", arr[i]);
    printf("\n");
}

double time_function(void (*func)(int[], int), int arr[], int n) {
    struct timespec begin, end;
    clock_gettime(CLOCK_MONOTONIC, &begin);

    func(arr, n); // Call the sorting function

    clock_gettime(CLOCK_MONOTONIC, &end);

    double time_spent = (end.tv_sec - begin.tv_sec) + (end.tv_nsec - begin.tv_nsec) / 1000000000.0;
    return time_spent;
}


//Main function to output runtimes to csv. 
int main(int argc, char *argv[]) {
    if (argc < 2) {
        printf("Usage: %s <max array size> [output file]\n", argv[0]);
        return 1;
    }

    int max_n = atoi(argv[1]);
    const char* output_file = argc > 2 ? argv[2] : "radix_sort_results.csv";

    FILE *file = fopen(output_file, "w");
    if (!file) {
        printf("Error opening file\n");
        return 1;
    }

    // Write CSV header
    fprintf(file, "Array Size,Time Taken (seconds),Memory Used (bytes)\n");

    for (int n = 1; n <= max_n; n++) {
        allocated_memory = 0; // Reset memory tracker for each iteration

        int *arr = (int *)malloc_track(n * sizeof(int));
        if (!arr) {
            printf("Memory allocation failed\n");
            fclose(file);
            return 1;
        }

        // Fill the array with random numbers
        srand(time(NULL));
        for (int i = 0; i < n; i++) {
            arr[i] = rand() % 1000;  // Random numbers from 0 to 999
        }

        double elapsed = time_function(radixSort, arr, n);
        //printf("Array Size: %d, Time Taken: %.6f seconds\n", n, elapsed);
        printf("Array Size: %d, Time Taken: %.6f seconds, Memory Used: %zu bytes\n", n, elapsed, allocated_memory);

        // Write to CSV
        fprintf(file, "%d,%.6f,%zu\n", n, elapsed, allocated_memory);

        free(arr);
    }

    fclose(file);
    return 0;
}
